src/audio/module_adapter/module/volume/volume_generic.c

```diff
@@ -52,50 +52,45 @@ static inline int32_t vol_mult_s24_to_s24(int32_t x, int32_t vol)
  * Copy and scale volume from 24/32 bit source buffer
  * to 24/32 bit destination buffer.
  */
-static void vol_s24_to_s24(struct comp_dev *dev, struct audio_stream *sink,
-			   const struct audio_stream *source, uint32_t frames)
+static void vol_s24_to_s24(struct processing_module *mod, struct input_stream_buffer *source,
+			   struct output_stream_buffer *sink, uint32_t frames, uint16_t nch)
 {
-	struct vol_data *cd = comp_get_drvdata(dev);
+	struct vol_data *cd = module_get_private_data(mod);
 	int32_t vol;
 	int32_t *x, *x0;
 	int32_t *y, *y0;
 	int nmax, n, i, j;
-	const int nch = source->channels;
 	int remaining_samples = frames * nch;
 #if CONFIG_COMP_PEAK_VOL
 	int32_t tmp = INT_MIN(32);
 #endif
 
-	x = source->r_ptr;
-	y = sink->w_ptr;
-	while (remaining_samples) {
-		nmax = VOL_BYTES_TO_S32_SAMPLES(audio_stream_bytes_without_wrap(source, x));
-		n = MIN(remaining_samples, nmax);
-		nmax = VOL_BYTES_TO_S32_SAMPLES(audio_stream_bytes_without_wrap(sink, y));
-		n = MIN(n, nmax);
-		for (j = 0; j < nch; j++) {
-			x0 = x + j;
-			y0 = y + j;
-			vol = cd->volume[j];
-			for (i = 0; i < n; i += nch) {
-				*y0 = vol_mult_s24_to_s24(*x0, vol);
+	x = (int32_t *)((uint8_t *)source->data + source->consumed);
```

Comment:
Could all the bytes be avoided in module API since they cause this kind of typecast burden? Pointers would be specific to data type int32_t, int16_t and offsets integer indexes. We don't have much components those operate with data bytes.

Reply:
@singalsu I have removed this now and retained the original logic. So I suppose the testbench verification doesnt need to be performed either.